finetune_llama2.py

```diff
@@ -0,0 +1,275 @@
+    optimizer: str = field(default="adam") 
+    auto_cast: bool = True
+    lazy_init: bool = True 
+    lr_scheduler: str = field(default="cosine") 
+    learning_rate: float = field(default=2e-5)
+    adam_beta1: float = field(default=0.9) 
+    adam_beta2: float = field(default=0.95)
+    weight_decay: float = field(default=0.1)
+    num_train_epochs: int = field(default=1) 
+    per_device_train_batch_size: int = field(default=1)
+    tp: int = field(default=4)
+    use_ddp: bool = True
+    model_max_length: int = field(
+        default=512,
+        metadata={"help": "Maximum sequence length. Sequences will be right padded (and possibly truncated)."},
+    )
+
+def tokenize_batch_for_finetune(batch, tokenizer: Optional[LlamaTokenizer] = None, max_length: int = 2048):
+    texts = [sample["prompt"] + sample["completion"] for sample in batch]
+    data = tokenizer(texts, return_tensors="pt", padding="max_length", truncation=True, max_length=max_length)
+    data = {k: v.cuda() for k, v in data.items()}
+    data["labels"] = data["input_ids"].clone()
+    return data
+
+def tokenize_batch_for_finetune_tp(batch, tokenizer: Optional[LlamaTokenizer] = None, max_length: int = 2048):
+    texts = batch["prompt"] + batch["completion"]
+    data = tokenizer(texts)
+    data["input_ids"] = torch.LongTensor(data["input_ids"])
+    data["attention_mask"] = torch.LongTensor(data["attention_mask"]) 
+    data["labels"] = deepcopy(data["input_ids"])
+    return data
+
+def print_rank0(msg):
+    rank = dist.get_rank()
+    if rank == 0:
+        print(msg)
+
+def get_model_size(model: torch.nn.Module):
+    total_numel = 0
+    for module in model.modules():
+        for p in module.parameters(recurse=False):
+            total_numel += p.numel()
+    return total_numel
+
+def get_tflops(grad_checkpoint, model_numel, batch_size, seq_len, step_time):
+    
+    return model_numel * batch_size * seq_len * (8 if grad_checkpoint else 6) / 1e12 / (step_time + 1e-12)
+
+def to_device(batch, device):
+    return {k: v.to(device) for k, v in batch.items()}
+
+def train():
+    parser = HfArgs((ModelArguments, DataArguments, TrainingArguments))
+    setup()
+    model_args, data_args, training_args = parser.parse_args_into_dataclasses()
+    # finetune loss function has already done in LlamaForCausalLM loss function.
+    model = LlamaForCausalLM.from_pretrained(
+        model_args.model_name_or_path,
+        cache_dir=training_args.cache_dir,
+    )
+    
+    """config = LlamaConfig(
+        hidden_size=768,
+        intermediate_size=1024,
+        num_hidden_layers=12,
+        num_attention_heads=8,
+        max_position_embeddings=1024,
+        num_key_value_heads=8
+    )
+    
+    model = LlamaForCausalLM(config)"""
+
+    # lazy_init
+    # torch DDP can't recognize the lazy_init, use it after DDP.
+    if not training_args.use_ddp and training_args.lazy_init:
+        lazy_init.replace_layers(model)
```

Comment:
I do not see this point of using `replace_layers` here. `model` is a model object that has been initialized already, replacing the layers will convert the initialized model to a lazy model, then this lazy model will get initialized again somewhere. This is the wrong way of using the lazy init modules.